cmd/kubeadm/app/master/selfhosted_etcd.go

```diff
@@ -0,0 +1,427 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package master
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/coreos/etcd-operator/pkg/spec"
+	"github.com/coreos/etcd-operator/pkg/util/k8sutil"
+	"github.com/coreos/etcd/clientv3"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api"
+	"k8s.io/client-go/pkg/api/v1"
+	ext "k8s.io/client-go/pkg/apis/extensions/v1beta1"
+	rbac "k8s.io/client-go/pkg/apis/rbac/v1beta1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	"k8s.io/kubernetes/cmd/kubeadm/app/images"
+)
+
+func launchEtcdOperator(cfg *kubeadmapi.MasterConfiguration, client *clientset.Clientset) error {
+	start := time.Now()
+
+	clusterRole := getEtcdClusterRole()
+	if _, err := client.RbacV1beta1().ClusterRoles().Create(&clusterRole); err != nil {
+		return fmt.Errorf("[self-hosted] Failed to create etcd-operator ClusterRole [%v]", err)
+	}
+
+	serviceAccount := getEtcdServiceAccount()
+	if _, err := client.CoreV1().ServiceAccounts(metav1.NamespaceSystem).Create(&serviceAccount); err != nil {
+		return fmt.Errorf("[self-hosted] Failed to create etcd-operator ServiceAccount [%v]", err)
+	}
+
+	clusterRoleBinding := getEtcdClusterRoleBinding()
+	if _, err := client.RbacV1beta1().ClusterRoleBindings().Create(&clusterRoleBinding); err != nil {
+		return fmt.Errorf("[self-hosted] Failed to create etcd-operator ClusterRoleBinding [%v]", err)
+	}
+
+	etcdOperatorDep := getEtcdOperatorDeployment(cfg)
+	if _, err := client.Extensions().Deployments(metav1.NamespaceSystem).Create(&etcdOperatorDep); err != nil {
+		return fmt.Errorf("[self-hosted] Failed to create etcd-operator deployment [%v]", err)
+	}
+
+	waitForPodsWithLabel(client, etcdOperator, true)
+	fmt.Printf("[self-hosted] etcd-operator deployment ready after %f seconds\n", time.Since(start).Seconds())
+
+	return nil
+}
+
+func CreateEtcdCluster(cfg *kubeadmapi.MasterConfiguration, client *clientset.Clientset) error {
+	start := time.Now()
+
+	// setup TPR client
+	restClient, err := getEtcdTPRClient()
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("[self-hosted] Waiting for etcd ThirdPartyResource to exist")
+	k8sutil.WaitEtcdTPRReady(restClient, time.Second*5, time.Minute*1, "kube-system")
+
+	seedPodIP, err := getBootEtcdPodIP(client)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("[self-hosted] Boot IP for etcd is %s\n", seedPodIP)
+
+	// TODO: Add etcd TLS. The etcd-operator accepts a series of secrets that are
```

Comment:
I take back my words on this.
Seems like there is a feature "ClusterFirstWithHostNet" that enable us to use service name in hostnet now.